{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b993d452",
   "metadata": {},
   "source": [
    "## SUMMARY\n",
    "\n",
    "Analysis of ligation products for spligation experiments.\n",
    "\n",
    "Briefly, for RT-PCR Nanopore data, each productâ€™s composition was confirmed by aligning reads to the corresponding amplicon with minimap2 (v2.28) and a custom junction file representing Csm cut sites. The abundance of each excision or ligation product was determined with HTSlib (v1.6)20, samtools (v1.6)21, and pysam (v0.18.0)22 and reported relative to total read depth within the crRNA target site.\n",
    "\n",
    "RNA-seq data (GEO accession number GSE220741) were analyzed as previously described in Colognori et al. 2023. Ligation products are assessed below using the bam files published in the original paper.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c0a2c4fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### IMPORT LIBS ####\n",
    "import pandas as pd\n",
    "from pandas.errors import EmptyDataError \n",
    "import numpy as np\n",
    "import shutil\n",
    "import os\n",
    "import sys\n",
    "import requests\n",
    "import subprocess\n",
    "import re\n",
    "import glob\n",
    "import pysamstats\n",
    "import pysam\n",
    "import Bio\n",
    "from Bio.Seq import reverse_complement\n",
    "from itertools import groupby\n",
    "from operator import itemgetter\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from collections import Counter\n",
    "\n",
    "## SUPRESS WARNINGS\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "## SLURM\n",
    "bash_header = [\"#!/bin/bash\", \"#SBATCH -p standard\", \"#SBATCH --job-name ALIGN_SPLIG\", \n",
    "               \"#SBATCH -o %j.out\", \"#SBATCH -e %j.err\", \"## ACTIVATE CONDA\", \n",
    "               'eval \"$(conda shell.bash hook)\"', 'conda activate Test_Py_3_9'\n",
    "              ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 838,
   "id": "95e33be0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## PROCESSING LBR: TACATCTACTAATGCTCTTCTGGCTTT\n",
      "TAC------TAATGCTCTTCTGGCTTT\n",
      "TAC------------TCTTCTGGCTTT\n",
      "TAC------------------GGCTTT\n",
      "TAC------------------------\n",
      "TACATCTAC------TCTTCTGGCTTT\n",
      "TACATCTAC------------GGCTTT\n",
      "TACATCTAC------------------\n",
      "TACATCTACTAATGC------GGCTTT\n",
      "TACATCTACTAATGC------------\n",
      "TACATCTACTAATGCTCTTCT------\n",
      "## PROCESSING NCL: AAGTTTGAATAGCTTCTGTCCCTCTGC\n",
      "AAG------TAGCTTCTGTCCCTCTGC\n",
      "AAG------------CTGTCCCTCTGC\n",
      "AAG------------------CTCTGC\n",
      "AAG------------------------\n",
      "AAGTTTGAA------CTGTCCCTCTGC\n",
      "AAGTTTGAA------------CTCTGC\n",
      "AAGTTTGAA------------------\n",
      "AAGTTTGAATAGCTT------CTCTGC\n",
      "AAGTTTGAATAGCTT------------\n",
      "AAGTTTGAATAGCTTCTGTCC------\n",
      "## PROCESSING NPM1: AAGTCTCTTTAAGAAAATAGTTTAAAC\n",
      "AAG------TAAGAAAATAGTTTAAAC\n",
      "AAG------------AATAGTTTAAAC\n",
      "AAG------------------TTAAAC\n",
      "AAG------------------------\n",
      "AAGTCTCTT------AATAGTTTAAAC\n",
      "AAGTCTCTT------------TTAAAC\n",
      "AAGTCTCTT------------------\n",
      "AAGTCTCTTTAAGAA------TTAAAC\n",
      "AAGTCTCTTTAAGAA------------\n",
      "AAGTCTCTTTAAGAAAATAGT------\n",
      "GCG------GTTCTGTAGCAGTGTTCAAGTGG\n",
      "GCG------------TAGCAGTGTTCAAGTGG\n",
      "GCG------------------TGTTCAAGTGG\n",
      "GCG------------------------AGTGG\n",
      "GCGGATCAA------TAGCAGTGTTCAAGTGG\n",
      "GCGGATCAA------------TGTTCAAGTGG\n",
      "GCGGATCAA------------------AGTGG\n",
      "GCGGATCAAGTTCTG------TGTTCAAGTGG\n",
      "GCGGATCAAGTTCTG------------AGTGG\n",
      "GCGGATCAAGTTCTGTAGCAG------AGTGG\n",
      "GCGGATCAAGTTCTGTAGCAGTGTTCAAGTGG\n",
      "GCG------------------------|GATCAAGTTCTGTCGCAGTGTTCAAGTGG\n",
      "GCG------------------------|------GTTCTGTCGCAGTGTTCAAGTGG\n",
      "GCG------------------------|------------TCGCAGTGTTCAAGTGG\n",
      "GCG------------------------|------------------TGTTCAAGTGG\n",
      "GCG------------------------|------------------------AGTGG\n",
      "GCGGATCAA------------------|GATCAAGTTCTGTCGCAGTGTTCAAGTGG\n",
      "GCGGATCAA------------------|------GTTCTGTCGCAGTGTTCAAGTGG\n",
      "GCGGATCAA------------------|------------TCGCAGTGTTCAAGTGG\n",
      "GCGGATCAA------------------|------------------TGTTCAAGTGG\n",
      "GCGGATCAA------------------|------------------------AGTGG\n",
      "GCGGATCAAGTTCTG------------|GATCAAGTTCTGTCGCAGTGTTCAAGTGG\n",
      "GCGGATCAAGTTCTG------------|------GTTCTGTCGCAGTGTTCAAGTGG\n",
      "GCGGATCAAGTTCTG------------|------------TCGCAGTGTTCAAGTGG\n",
      "GCGGATCAAGTTCTG------------|------------------TGTTCAAGTGG\n",
      "GCGGATCAAGTTCTG------------|------------------------AGTGG\n",
      "GCGGATCAAGTTCTGTCGCAG------|GATCAAGTTCTGTCGCAGTGTTCAAGTGG\n",
      "GCGGATCAAGTTCTGTCGCAG------|------GTTCTGTCGCAGTGTTCAAGTGG\n",
      "GCGGATCAAGTTCTGTCGCAG------|------------TCGCAGTGTTCAAGTGG\n",
      "GCGGATCAAGTTCTGTCGCAG------|------------------TGTTCAAGTGG\n",
      "GCGGATCAAGTTCTGTCGCAG------|------------------------AGTGG\n",
      "GCGGATCAAGTTCTGTCGCAGTGTTCA|GATCAAGTTCTGTCGCAGTGTTCAAGTGG\n",
      "GCGGATCAAGTTCTGTCGCAGTGTTCA|------GTTCTGTCGCAGTGTTCAAGTGG\n",
      "GCGGATCAAGTTCTGTCGCAGTGTTCA|------------TCGCAGTGTTCAAGTGG\n",
      "GCGGATCAAGTTCTGTCGCAGTGTTCA|------------------TGTTCAAGTGG\n",
      "GCGGATCAAGTTCTGTCGCAGTGTTCA|------------------------AGTGG\n",
      "##XIST_crNA_1\n",
      "# TARGET SITE:  CGGATCCAGTTCTGTCGCAGTGTTCAAGTGGC\n",
      "##XIST_crNA_2\n",
      "# TARGET SITE:  CATGGCGGGCTGTGCTTTGTTAGGTTGTCCAA\n"
     ]
    }
   ],
   "source": [
    "#### SETUP ####\n",
    "'''\n",
    "Configure strings corresponding to the crRNA target site.\n",
    "Strings for ligation/deletion products will be used for display\n",
    "'''\n",
    "\n",
    "pwd = os.getcwd()\n",
    "analysis_key = \"Spligation_Key_12162025.xlsx\"\n",
    "sup_table_df = pd.read_excel(\"Spligation_Key_12162025.xlsx\", sheet_name=\"Sup_Table_ROIs\")\n",
    "\n",
    "cut_indices = [3, 9, 15, 21] # Cut site locations\n",
    "deltas = [24, 18, 12, 6] # Deletion sizes\n",
    "\n",
    "# Target site strings\n",
    "target_dict = {\"LBR\":[\"TACATCTACTAATGCTCTTCTGGCTTT\", 27], # [String, length]\n",
    "               \"NCL\":[\"AAGTTTGAATAGCTTCTGTCCCTCTGC\", 27],\n",
    "               \"NPM1\":[\"AAGTCTCTTTAAGAAAATAGTTTAAAC\", 27]\n",
    "}\n",
    "\n",
    "# REF ROI COORDs\n",
    "ref_roi_coords = pd.read_excel(\"Spligation_Key_12162025.xlsx\", \n",
    "                               sheet_name = \"Ref_ROI_Locations\"\n",
    "                              )\n",
    "ligation_str_dict = {}\n",
    "\n",
    "# ENDOGENOUS\n",
    "for row in ref_roi_coords.itertuples():\n",
    "    ref = row.Construct\n",
    "    transcript = ref.split(\"_\")[0]\n",
    "    target_region = target_dict[transcript][0]\n",
    "    print(f\"## PROCESSING {transcript}: {target_region}\")\n",
    "    \n",
    "    # LIGATION STRINGS\n",
    "    cp_ends = [\"\".join(target_region[:n]) for n in cut_indices]\n",
    "    oh_ends = [\"\".join(target_region[n:]) for n in cut_indices]+['']\n",
    "    ligation_strs = [cp_end+(27-len(cp_end)-len(oh_end))*\"-\"+oh_end for cp_end in cp_ends for oh_end in oh_ends]\n",
    "    ligation_strs = [x for x in ligation_strs if len(x)==27 and \"-\" in x]\n",
    "    print('\\n'.join(ligation_strs))\n",
    "    \n",
    "    ligation_str_dict[transcript] = ligation_strs\n",
    "    \n",
    "# 1 CUT Cis: 'Cis_1Cut'\n",
    "reporter_target_region = 'GCGGATCAAGTTCTGTAGCAGTGTTCAAGTGG'\n",
    "cut_indices = [3, 9, 15, 21, 27] # Plot full target region (32nt)\n",
    "cp_ends = [\"\".join(reporter_target_region[:n]) for n in cut_indices]\n",
    "oh_ends = [\"\".join(reporter_target_region[n:]) for n in cut_indices]\n",
    "ligation_strs = [cp_end+(32-len(cp_end)-len(oh_end))*\"-\"+oh_end for cp_end in cp_ends for oh_end in oh_ends]\n",
    "ligation_strs = [x for x in ligation_strs if len(x)==32 and \"-\" in x]+[reporter_target_region]\n",
    "print('\\n'.join(ligation_strs))\n",
    "ligation_str_dict['Cis_1Cut'] = ligation_strs\n",
    "\n",
    "# 2 CUT Cis: '2Cut_Short_Cis_Amplicon', '2Cut_Long_Cis_Amplicon'\n",
    "large_del_conversion_dict = {}\n",
    "n_term = \"GCGGATCAAGTTCTGTCGCAGTGTTCAAGTGG\"\n",
    "cp_ends = [\"\".join(n_term[:n]) for n in cut_indices]\n",
    "oh_ends = [\"\".join(n_term[n:]) for n in cut_indices]\n",
    "formatted_large_del_ligation_strs = [cp_end+(27-len(cp_end))*\"-\"+\"-----|\"\\\n",
    "                 +\"---\"\\\n",
    "                 +(29-len(oh_end))*\"-\"+oh_end for cp_end in cp_ends for oh_end in oh_ends]\n",
    "\n",
    "print('\\n'.join(large_del_ligation_strs))\n",
    "large_del_conditions = ['Cis_2CutShort', 'Cis_2CutLong']\n",
    "\n",
    "for large_del_condition in large_del_conditions:\n",
    "    del_row = sup_table_df.loc[sup_table_df.Analysis == large_del_condition]\n",
    "    del_start = int(del_row.ROI_Skip_Start.values[0])\n",
    "    del_stop = int(del_row.ROI_Skip_Stop.values[0])\n",
    "    del_str = (del_stop-del_start)*\"-\"\n",
    "    large_del_ligation_strs = [cp_end+(27-len(cp_end))*\"-\"+\"|\"\\\n",
    "                 +(29-len(oh_end))*\"-\"+oh_end for cp_end in cp_ends for oh_end in oh_ends]\n",
    "    temp_lig_strs = [x.replace('|', f'{del_str}') for x in large_del_ligation_strs]\n",
    "    ligation_str_dict[large_del_condition] = temp_lig_strs\n",
    "    large_del_conversion_dict[large_del_condition] = dict(zip(temp_lig_strs,formatted_large_del_ligation_strs))\n",
    "\n",
    "# TRANS\n",
    "n_term = \"GCGGATCAAGTTCTGTCGCAGTGTTCAAGTGG\"\n",
    "cp_ends = [\"\".join(n_term[:n]) for n in cut_indices]\n",
    "oh_ends = [\"\".join(n_term[n:]) for n in cut_indices]\n",
    "trans_del_ligation_strs = [cp_end+(27-len(cp_end))*\"-\"+\n",
    "                 +(29-len(oh_end))*\"-\"+oh_end for cp_end in cp_ends for oh_end in oh_ends]\n",
    "clean_pipe = [x.replace('|', f'') for x in trans_del_ligation_strs]\n",
    "ligation_str_dict['Trans_Basic'] = clean_pipe\n",
    "trans_conversion_dict = dict(zip(clean_pipe,formatted_large_del_ligation_strs))\n",
    "\n",
    "# XIST\n",
    "xist_conditions = ['XIST_crNA_1', 'XIST_crNA_2']\n",
    "xist_conversion_dict = {}\n",
    "for xist_condition in xist_conditions:\n",
    "    print(f\"##{xist_condition}\")\n",
    "    xist_row = sup_table_df.loc[sup_table_df.Analysis == xist_condition]\n",
    "    crrna_str = xist_row.ROI_Str.values[0]\n",
    "    target_site = str(Bio.Seq.Seq(crrna_str).reverse_complement())\n",
    "    \n",
    "    print(\"# TARGET SITE: \",target_site)\n",
    "    cut_indices = [3, 9, 15, 21]\n",
    "    deltas = [24, 18, 12, 6]\n",
    "\n",
    "    cut_indices = [3, 9, 15, 21, 27] # Plot full target region (32nt)\n",
    "    cp_ends = [\"\".join(target_site[:n]) for n in cut_indices]\n",
    "    oh_ends = [\"\".join(target_site[n:]) for n in cut_indices]\n",
    "    ligation_strs = [cp_end+(32-len(cp_end)-len(oh_end))*\"-\"+oh_end for cp_end in cp_ends for oh_end in oh_ends]\n",
    "    ligation_strs = [x for x in ligation_strs if len(x)==32 and \"-\" in x]+[target_site]\n",
    "    ligation_str_dict[xist_condition] = ligation_strs\n",
    "    \n",
    "    conversion_strs = [str(Bio.Seq.Seq(x).reverse_complement()) for x in ligation_strs]\n",
    "    xist_conversion_dict[xist_condition] = dict(zip(conversion_strs,ligation_strs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 840,
   "id": "4a757205",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LBR_Amplicon_With_Target_Site.fasta TACATCTACTAATGCTCTTCTGGCTTT ['TAC', 'TACATCTAC', 'TACATCTACTAATGC', 'TACATCTACTAATGCTCTTCT', 'TACATCTACTAATGCTCTTCTGGCTTT']\n",
      "[3, 9, 15, 21, 27] [24, 18, 12, 6, 0, 0] !!!!14\n",
      "TAC------TAATGCTCTTCTGGCTTT: 27, 466__466\n",
      "TAC------------TCTTCTGGCTTT: 27, 466__472\n",
      "TAC------------------GGCTTT: 27, 466__478\n",
      "TAC------------------------: 27, 466__484\n",
      "TAC------------------------: 27, 466__490\n",
      "TACATCTAC------TCTTCTGGCTTT: 27, 466__490\n",
      "TACATCTAC------------GGCTTT: 27, 472__466\n",
      "TACATCTAC------------------: 27, 472__472\n",
      "TACATCTAC------------------: 27, 472__478\n",
      "TACATCTACTAATGC------GGCTTT: 27, 472__484\n",
      "TACATCTACTAATGC------------: 27, 472__490\n",
      "TACATCTACTAATGC------------: 27, 472__490\n",
      "TACATCTACTAATGCTCTTCT------: 27, 478__466\n",
      "TACATCTACTAATGCTCTTCT------: 27, 478__472\n",
      "NCL_Amplicon_With_Target_Site.fasta AAGTTTGAATAGCTTCTGTCCCTCTGC ['AAG', 'AAGTTTGAA', 'AAGTTTGAATAGCTT', 'AAGTTTGAATAGCTTCTGTCC', 'AAGTTTGAATAGCTTCTGTCCCTCTGC']\n",
      "[3, 9, 15, 21, 27] [24, 18, 12, 6, 0, 0] !!!!14\n",
      "AAG------TAGCTTCTGTCCCTCTGC: 27, 274__274\n",
      "AAG------------CTGTCCCTCTGC: 27, 274__280\n",
      "AAG------------------CTCTGC: 27, 274__286\n",
      "AAG------------------------: 27, 274__292\n",
      "AAG------------------------: 27, 274__298\n",
      "AAGTTTGAA------CTGTCCCTCTGC: 27, 274__298\n",
      "AAGTTTGAA------------CTCTGC: 27, 280__274\n",
      "AAGTTTGAA------------------: 27, 280__280\n",
      "AAGTTTGAA------------------: 27, 280__286\n",
      "AAGTTTGAATAGCTT------CTCTGC: 27, 280__292\n",
      "AAGTTTGAATAGCTT------------: 27, 280__298\n",
      "AAGTTTGAATAGCTT------------: 27, 280__298\n",
      "AAGTTTGAATAGCTTCTGTCC------: 27, 286__274\n",
      "AAGTTTGAATAGCTTCTGTCC------: 27, 286__280\n",
      "NPM1_Amplicon_With_Target_Site.fasta AAGTCTCTTTAAGAAAATAGTTTAAAC ['AAG', 'AAGTCTCTT', 'AAGTCTCTTTAAGAA', 'AAGTCTCTTTAAGAAAATAGT', 'AAGTCTCTTTAAGAAAATAGTTTAAAC']\n",
      "[3, 9, 15, 21, 27] [24, 18, 12, 6, 0, 0] !!!!14\n",
      "AAG------TAAGAAAATAGTTTAAAC: 27, 443__443\n",
      "AAG------------AATAGTTTAAAC: 27, 443__449\n",
      "AAG------------------TTAAAC: 27, 443__455\n",
      "AAG------------------------: 27, 443__461\n",
      "AAG------------------------: 27, 443__467\n",
      "AAGTCTCTT------AATAGTTTAAAC: 27, 443__467\n",
      "AAGTCTCTT------------TTAAAC: 27, 449__443\n",
      "AAGTCTCTT------------------: 27, 449__449\n",
      "AAGTCTCTT------------------: 27, 449__455\n",
      "AAGTCTCTTTAAGAA------TTAAAC: 27, 449__461\n",
      "AAGTCTCTTTAAGAA------------: 27, 449__467\n",
      "AAGTCTCTTTAAGAA------------: 27, 449__467\n",
      "AAGTCTCTTTAAGAAAATAGT------: 27, 455__443\n",
      "AAGTCTCTTTAAGAAAATAGT------: 27, 455__449\n"
     ]
    }
   ],
   "source": [
    "#### PROCESS REFERENCES ####\n",
    "'''\n",
    "Create splice junction file corresponding to 6nt cut sites characteristic of CRISPR-Csm\n",
    "'''\n",
    "\n",
    "contig_junction_dict = {}\n",
    "start_stop_str_dicts = {}\n",
    "\n",
    "for idx,roi in ref_roi_coords.iterrows():\n",
    "    construct = roi['Construct']\n",
    "    target_starts = [int(x) for x in roi[['Target1_0Based_Start', 'Target2_0Based_Start']].values if str(x) != \"nan\"]\n",
    "\n",
    "    seq = target_dict[construct.split('_')[0]][0]\n",
    "    max_str_len = target_dict[construct.split('_')[0]][-1]\n",
    "\n",
    "    \n",
    "    # LIGATION STRINGS\n",
    "    cp_ends = [\"\".join(seq[:n]) for n in cut_indices]\n",
    "    oh_ends = [\"\".join(seq[n:]) for n in cut_indices]+['']\n",
    "    len_cp_ends = [len(cp_end) for cp_end in cp_ends]\n",
    "    len_oh_ends = [len(oh_end) for oh_end in oh_ends]\n",
    "    ligation_strs = [cp_end+(27-len(cp_end)-len(oh_end))*\"-\"+oh_end for cp_end in cp_ends for oh_end in oh_ends]\n",
    "    ligation_strs = [x for x in ligation_strs if len(x)==27 and \"-\" in x]\n",
    "    print(construct,seq, cp_ends)\n",
    "    \n",
    "    # DEFINE JUNCTION SITES\n",
    "    ligation_locs = [f'{len(cp_end)+target_starts[0]}__{len(cp_end)+target_starts[0]+(max_str_len-len(cp_end)-len(oh_end))}' for cp_end in cp_ends for oh_end in oh_ends]\n",
    "    start_stop_str_dict = dict(zip(ligation_locs, ligation_strs))\n",
    "    start_stop_str_dicts[construct] = start_stop_str_dict\n",
    "    print(len_cp_ends, len_oh_ends, f\"!!!!{len(ligation_strs)}\")\n",
    "    print(\"\\n\".join([f\"{ligation_str}: {len(ligation_str)}, {ligation_locs[idx]}\" for idx,ligation_str in enumerate(ligation_strs)]))\n",
    "\n",
    "    # SAVE BED6\n",
    "    bed6_lines = []\n",
    "    strand = \"+\"\n",
    "    temp_dict = {}\n",
    "    \n",
    "    for i, loc in enumerate(ligation_locs):\n",
    "        start, end = loc.split(\"__\")\n",
    "        bed6_lines.append(f\"{construct.replace('.fasta','')}\\t{start}\\t{end}\\tjunc{i+1}\\t0\\t{strand}\")\n",
    "        temp_dict[f\"junc{i+1}\"] = loc\n",
    "    contig_junction_dict[construct]= temp_dict\n",
    "\n",
    "    # Save to file\n",
    "    with open(f\"References/{construct.replace('.fasta','')}_junctions.bed\", \"w\") as f:\n",
    "        for line in bed6_lines:\n",
    "            f.write(line + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 841,
   "id": "b604fddb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## PROCESSING: LBR__D9S6PX_1_EndoDeluxe-LBR\n",
      "/groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Data/D9S6PX_1_EndoDeluxe-LBR.fastq\n",
      "/groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/References/LBR_Amplicon_With_Target_Site.fasta\n",
      "## PROCESSING: NCL__D9S6PX_2_EndoDeluxe-NCL\n",
      "/groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Data/D9S6PX_2_EndoDeluxe-NCL.fastq\n",
      "/groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/References/NCL_Amplicon_With_Target_Site.fasta\n",
      "## PROCESSING: NPM1__D9S6PX_3_EndoDeluxe-NPM1\n",
      "/groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Data/D9S6PX_3_EndoDeluxe-NPM1.fastq\n",
      "/groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/References/NPM1_Amplicon_With_Target_Site.fasta\n"
     ]
    }
   ],
   "source": [
    "#### SETUP ALIGNMENTS ####\n",
    "'''\n",
    "Align Nanopore data to amplicon from RT-PCR experiments\n",
    "'''\n",
    "\n",
    "threads = \"$SLURM_CPUS_ON_NODE\"\n",
    "bash_lines = bash_header\n",
    "bash_script_name = \"Nanopore_Alignments.sh\"\n",
    "\n",
    "alignment_key = pd.read_excel(\"Spligation_Key_12162025.xlsx\", sheet_name=\"Alignment_Key\")\n",
    "\n",
    "\n",
    "data_dir = f\"{pwd}/Data/\"\n",
    "ref_dir = f\"{pwd}/References/\"\n",
    "alignment_dir = f\"{pwd}/Alignments/\"\n",
    "\n",
    "fastas = glob.glob(f\"References/*.fasta\")\n",
    "juncs = glob.glob(f\"References/*junctions.bed\")\n",
    "fastqs = glob.glob(f\"{pwd}/Data/*.fastq\")\n",
    "\n",
    "qcd = []\n",
    "bams_1ref_basic = {}\n",
    "bams_1ref_splice = {}\n",
    "for idx,row in alignment_key.iterrows(): \n",
    "    analysis_id = row['Sample_Description'].replace(' ', \"_\")\n",
    "    fq = f'{data_dir}{row[\"FASTQ_Name\"]}'\n",
    "    sample_id = f\"{analysis_id}__\" + fq.split(\"/\")[-1].split(\".\")[0]\n",
    "    ref_1seq = f\"{ref_dir}{row.Reference_for_Alignment}.fasta\"\n",
    "    \n",
    "    print(f\"## PROCESSING: {sample_id}\")\n",
    "    print(fq)\n",
    "    print(ref_1seq)\n",
    "    \n",
    "\n",
    "    bash_lines.append(f'#### {analysis_id}')\n",
    "    \n",
    "    # QC\n",
    "    qc_dir = data_dir\n",
    "    if fq not in qcd:\n",
    "        bash_lines.append(\"# QC\")\n",
    "        qc_str = f\"NanoPlot -t {threads} --fastq {fq} --maxlength 10000 --plots dot --prefix {sample_id} --outdir {qc_dir}\"\n",
    "        bash_lines.append(qc_str)\n",
    "        qcd.append(fq)\n",
    "        \n",
    "    \n",
    "    # 1REF SPLICE MODE ALIGNMENT\n",
    "    bash_lines.append(\"# 1REF SPLICE MODE ALIGNMENT\")\n",
    "    bam_1ref_splice = f\"{alignment_dir}{sample_id}__1ref_splice_sorted.bam\"\n",
    "    bams_1ref_splice[sample_id] = bam_1ref_splice\n",
    "    align_1ref_splice = (f\"minimap2 -ax splice -k12 -t {threads} -J 1 {ref_1seq} {fq} \"\n",
    "                         f\"| samtools view -bS -@ {threads} - \"\n",
    "                         f\"| samtools sort -@ {threads} -o {bam_1ref_splice} - && \"\n",
    "                         f\"samtools index  -@ {threads} {bam_1ref_splice}\"\n",
    "                        )\n",
    "    bash_lines.append(align_1ref_splice)\n",
    "\n",
    "with open(bash_script_name, \"w\") as f:\n",
    "    f.writelines(\"\\n\".join(bash_lines))\n",
    "\n",
    "\n",
    "#!sbatch Nanopore_Alignments.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 842,
   "id": "6d1a1bb0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LBR__D9S6PX_1_EndoDeluxe-LBR\n",
      "bamCoverage -b /groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Alignments/LBR__D9S6PX_1_EndoDeluxe-LBR__1ref_splice_sorted.bam -o /groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Alignments/LBR__D9S6PX_1_EndoDeluxe-LBR__1ref_splice_sorted.bw --numberOfProcessors 16 --binSize 1 bamCoverage -b /groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Alignments/LBR__D9S6PX_1_EndoDeluxe-LBR__1ref_splice_sorted.bam -o /groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Alignments/LBR__D9S6PX_1_EndoDeluxe-LBR__1ref_splice_sorted_CPM.bw --normalizeUsing CPM --numberOfProcessors 16 --binSize 1\n",
      "NCL__D9S6PX_2_EndoDeluxe-NCL\n",
      "bamCoverage -b /groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Alignments/NCL__D9S6PX_2_EndoDeluxe-NCL__1ref_splice_sorted.bam -o /groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Alignments/NCL__D9S6PX_2_EndoDeluxe-NCL__1ref_splice_sorted.bw --numberOfProcessors 16 --binSize 1 bamCoverage -b /groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Alignments/NCL__D9S6PX_2_EndoDeluxe-NCL__1ref_splice_sorted.bam -o /groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Alignments/NCL__D9S6PX_2_EndoDeluxe-NCL__1ref_splice_sorted_CPM.bw --normalizeUsing CPM --numberOfProcessors 16 --binSize 1\n",
      "NPM1__D9S6PX_3_EndoDeluxe-NPM1\n",
      "bamCoverage -b /groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Alignments/NPM1__D9S6PX_3_EndoDeluxe-NPM1__1ref_splice_sorted.bam -o /groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Alignments/NPM1__D9S6PX_3_EndoDeluxe-NPM1__1ref_splice_sorted.bw --numberOfProcessors 16 --binSize 1 bamCoverage -b /groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Alignments/NPM1__D9S6PX_3_EndoDeluxe-NPM1__1ref_splice_sorted.bam -o /groups/doudna/projects/mtrinidad_projects/Csm_RNASeq_David/Spligation/Nanopore_12182025/Alignments/NPM1__D9S6PX_3_EndoDeluxe-NPM1__1ref_splice_sorted_CPM.bw --normalizeUsing CPM --numberOfProcessors 16 --binSize 1\n"
     ]
    }
   ],
   "source": [
    "#### SETUP BIGWIG COVERAGE TRACKS\n",
    "'''\n",
    "Create bigwigs for IGV coverage tracks\n",
    "'''\n",
    "threads = \"$SLURM_CPUS_ON_NODE\"\n",
    "bash_lines = bash_header\n",
    "bash_script_name = \"Bigwigs.sh\"\n",
    "\n",
    "for sample_id, bam in bams_1ref_splice.items():  \n",
    "    print(sample_id)\n",
    "    ## BIGWIG\n",
    "    bash_lines.append(f\"# BigWig\")\n",
    "    bw_str = f\"bamCoverage -b {bam} -o {bam.replace('.bam', '.bw')} --numberOfProcessors {threads} --binSize 1\"\n",
    "    bash_lines.append(bw_str)\n",
    "    \n",
    "    bash_lines.append(f\"# CPM BigWig\")\n",
    "    bw_cpm_str = f\"bamCoverage -b {bam} -o {bam.replace('.bam', '_CPM.bw')} --normalizeUsing CPM --numberOfProcessors {threads} --binSize 1\"\n",
    "    bash_lines.append(bw_cpm_str)\n",
    "    print(bw_str, bw_cpm_str)\n",
    "\n",
    "with open(bash_script_name, \"w\") as f:\n",
    "    f.writelines(\"\\n\".join(bash_lines))\n",
    "\n",
    "#!sbatch Bigwigs.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 843,
   "id": "d104233b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#### Processing LBR\n",
      "3153 TACATCTAC------------------\n",
      "549 TACATCTACTAATGCTCTTC------T\n",
      "14 CCGCCATGGCCGAGGAAGGCATTGCTGCTGGAGGTGT\n",
      "10 ---ATGT-CTCCCCCTCCTTCTAGG--GG\n",
      "9 ---------GTGGTGGCCGGGGCATGGACCGAGGTGGCTTTGGT\n",
      "9 GACATCTAC------------------\n",
      "5 TACATCTA-------------------\n",
      "3 TAC--GTAC------------------\n",
      "3 TCACATCTAC------------------\n",
      "3 TACATCGACGAATGCTCTTC------T\n",
      "2 TACATCTACTAATGCTC--------TT\n",
      "2 TACATC--------------------T\n",
      "2 TACATCTAC--------------C---\n",
      "2 ----TGTAGCCTGTTGCCCATCCCCAGG---G\n",
      "2 TACATCTACCAA----------G----\n",
      "2 TACGTCT-CTAATGCTCTTC------T\n",
      "2 TACATC---------------------\n",
      "2 TACAT-TAC------------------\n",
      "2 TACA----------------------T\n",
      "2 -ACATCTAC------------------\n",
      "2 GACATC----AA---------------\n",
      "1 TACATATACTA------------C---\n",
      "1 TACAT----------------------\n",
      "1 TACATCTACCAA---------GG----\n",
      "1 ---------TCCATGTCTCCCCTTGTCAGG--GG\n",
      "1 TACATCTGCCAA----------G----\n",
      "1 TACATTTATTAATGCTCTTC------T\n",
      "1 TAC--CTACGTA-----CTT---AC---\n",
      "1 TACATCTAC----------C-GG----\n",
      "1 TACATCTCTTAATGCTCTTC------T\n",
      "1 TA-----AC----------C---C---\n",
      "1 TACATCTACTAATGCTCTT---AC-TT\n",
      "1 TACGT---------CTCTT-T---TAC\n",
      "1 GGCATTGCTGCTGGAGGTGTAATGG----\n",
      "1 TACGTCT----------------C---\n",
      "1 TAC-TCTACTAATGCTCTTC------T\n",
      "1 TACATCTACTGATGCT-TACT--CT--\n",
      "1 TACA--TAC--A---TCTAC-------\n",
      "1 TACATCTACTAATGCTCTTCTGGC---\n",
      "1 TACAT---CTAATGCTCTTC------T\n",
      "1 -------AGTA---CTCTTC-------\n",
      "1 TACATCTACTTAATGCTCTTC------T\n",
      "1 -----------GTGCGTCTT-----TAC\n",
      "1 TACGTCTAC------------------\n",
      "1 TACATCT--------------------\n",
      "1 ---ATCCA-TATCTGCCTCCTTCTAGG--GG\n",
      "1 TAC----------GCATCTT-----TAC\n",
      "1 TACAT-------TG---TTC---C---\n",
      "1 -----CGACTA-----CTT---AC---\n",
      "1 --------CTAATGCTCTTC------T\n",
      "1 ---------GTGGTGGCCGGGGCATGGACCGAGGTGGCTTT\n",
      "1 TACACATCTAC------------------\n",
      "1 TACAT----------TCTAC-------\n",
      "1 TACATCT----------------C---\n",
      "1 ---------------------------\n",
      "1 TA----TAC--A---TCTTC---C---\n",
      "1 TACATCT-CTA-----------G----\n",
      "1 ----TTT--TTTTTTTTTTTT---TTT\n",
      "1 GGCACCCCCCAA-----------C---\n",
      "1 --CATCTACTAATGCTCTTC------T\n",
      "1 AACATCTAC------------------\n",
      "1 T--ATCTACTAATGCTCTTC------T\n",
      "1 GA-A--TACTA------------C---\n",
      "1 GAGA----------------TGAC---\n",
      "1 GACACTC-ACTAATGCTCTTC------T\n",
      "1 TACATCTACTAATGCT-TTTTG-----\n",
      "1 T-CATCCAC--A-------------TT\n",
      "1 -GCAGCCCACCAATGCTCTTC------T\n",
      "1 GACA----------CTC-----AC---\n",
      "1 GTACATCTAC------------------\n",
      "1 GCACATCTACTAATGCTCTTC------T\n",
      "1 TACATCTACGAACGC-CTTC-----TT\n",
      "1 GACA-CGAC------------------\n",
      "1 TACATCTAC-------------G----\n",
      "1 TACAT-------TGAAC----------\n",
      "1 TACATCGACGAATGCTC--------TT\n",
      "1 TACATCT-CTATGAATGCTCTTC------T\n",
      "1 TACATC----------------G---T\n",
      "1 TAGA-CTAC------------------\n",
      "1 GAATATCGACTGAACGC-CTTCTCG----\n",
      "1 TACATCTACCAAGG----ACTGG--TT\n",
      "1 GAC-TC--CCA--------CAGG---T\n",
      "1 TACATTTA-TAGTG------------T\n",
      "1 TACACTCCAACGAACGCTCTTC------T\n",
      "1 TAGATCTACTAATGCTCTTC------T\n",
      "1 GCAACATC-ATGAATGCTCTTC------T\n",
      "1 TGCTTGTAC-------------G----\n",
      "1 GACATCTACTAATGCTCTTC------T\n",
      "1 TACATCTAACTA----TTTTCT---TCT\n",
      "1 TACAT----TA---------TG-----\n",
      "1 GACACTCAACTAATGCTCTTC------T\n",
      "1 GACA------------------GATTA\n",
      "1 CCGCCATGGCCGAGGAAAGCATCGCTGCTGGAGGTGT\n",
      "1 TACATCTACCA----------------\n",
      "1 -AGATCTAC------------------\n",
      "1 TACAATCTAC------------------\n",
      "1 GACATCTCTACTAATGCTCTTC------T\n",
      "1 -AC-T-TAC------------------\n",
      "1 TACATCTACCAAT--------------\n",
      "1 TACATCTACTAATGCTCTTCTG-----\n",
      "1 -ATATCTAC------------------\n",
      "1 GA------------------TGAC---\n",
      "1 TACATCTACTGATGCTCTT---G---T\n",
      "1 T-CGCCGAGGAAGGCATTGCTGCTGGAGGTGT\n",
      "1 TA----TACACGTGCATCTTC---C--C\n",
      "1 CACAT-T-CCAA-----------C---\n",
      "1 TACAT------A---------------\n",
      "1 TACATCTACGAATGCTCTTC------T\n",
      "1 TACATCTAC----------C-GG---T\n",
      "1 TACATCTACTAATGCTGCTT---G-TCT\n",
      "1 TACATCTACTAATGCTCTTCT--C---\n",
      "1 TAGCACATCTAC------------------\n",
      "1 TACATCTACTAGTGCTCTTC------T\n",
      "1 --GATCTCCTA------------C---\n",
      "1 TACATCTCCCA-----------G----\n",
      "1 TACATC------TGCAC----------\n",
      "1 TACA--TAC------------------\n",
      "1 GA---CTAC------------------\n",
      "1 TACATCTACCAA---------GGAT--\n",
      "1 TAGAT-TCCCA------------C---\n",
      "1 TACA-------------TTC-------\n",
      "#### Processing NCL\n",
      "2956 AAGTTTGAA------------------\n",
      "967 AAGTTTGAATAGCTT------------\n",
      "181 AAGTTTGA-------------------\n",
      "138 ---------------------------\n",
      "91 AAGTTTGAAT----------------C\n",
      "88 AAGTTTGAA-ACGGAAACGGAAAGCAGAGGG\n",
      "73 AAG------------------------\n",
      "62 AGGTTTGA-------------------\n",
      "20 ------------CCTC---CCCACTGGG\n",
      "10 AAGTTTGAA----------------G-\n",
      "7 AAG-TTGAA------------------\n",
      "4 TAGAT--CACAGCTACTTTTCTTGTCCCTGTGGGG\n",
      "3 AAGTT----------------------\n",
      "3 AAATTTGAA------------------\n",
      "3 AAG-TTGAATAGCTT------------\n",
      "3 -AGTTTGAA------------------\n",
      "2 AAG-------AG---CTG-------AA\n",
      "2 ------------ATTCCATCGATCTGTC\n",
      "2 AGTTTTGAA------------------\n",
      "2 AAGATTTGAA------------------\n",
      "2 AAGATTTGAATAGCTT------------\n",
      "2 AAGTTTG-ATAGCTT------------\n",
      "2 AAGTTTGAATAGCT----------TG-\n",
      "2 AAGTTTGAA--------G-------G-\n",
      "2 AAGTTTGAATAGCTT-----------C\n",
      "2 GAGTTTGAA------------------\n",
      "2 AAGTTTGA-----------------GC\n",
      "2 AA-------------------------\n",
      "1 AAGTTTGAACGAG------------T--\n",
      "1 AAATTTGAGAA----------------\n",
      "1 -AG---GA-----TT---TCC----AA\n",
      "1 AAGTTTGAATAGCTTCAG---------\n",
      "1 AAGTTTGAATAGCTTC----------A\n",
      "1 AAG-CTGAATAGCTT------------\n",
      "1 -AGTGATGAATAGCTT------------\n",
      "1 AAAGATTGAAAA---TGGGT--GTAAG-\n",
      "1 AAGTTTGAATAGCT--TG-------GA\n",
      "1 GAGTTTGAATAGCTT------------\n",
      "1 AAGATTTAAACAACT-------------\n",
      "1 AAGTTTGAA-AGC--------------\n",
      "1 AAGTTTGGA-A----------------\n",
      "1 AAGTTTGAA-ATCAT---------T--\n",
      "1 AAGTT--AAAA--------------AC\n",
      "1 ---------------CAGTC-------\n",
      "1 AAGTTCG-ATGGCT----------AG-\n",
      "1 A-----GAA------------------\n",
      "1 AAAGAGTTTGAA------------------\n",
      "1 AAAGTTTGAA------------------\n",
      "1 CAG------------------CTC---\n",
      "1 AAGTTTGGATGGCT-------------\n",
      "1 AAGTTTGAAACGGAAGAGGAAG--CAGAGGG\n",
      "1 ACGTCTGAA------------------\n",
      "1 ------------C--CTGTCCACT-GGG\n",
      "1 AAGTTTGAA-ACGGAAACGGAAAGCAGAGG\n",
      "1 AAGTTTGAATAGCT----------AG-\n",
      "1 AAG-ACGAATAGCTT------------\n",
      "1 AA-CTTGAA------------------\n",
      "1 AAGTTTGAATAGCTTC---------AC\n",
      "1 AAGATTTGAATACCTT------------\n",
      "1 AAGTTTGAATAGCT--TG-----AAG-\n",
      "1 CAGT-----------------------\n",
      "1 -A----GAA------------------\n",
      "1 AAGTTTGAATA---------CCT----\n",
      "1 A-----GAA-A--------------GA\n",
      "1 ------------CCTC---CCCACTGGA\n",
      "1 AAGTTTGAGAA----------------\n",
      "1 AAG-------AGC------------AA\n",
      "1 CATTCTCTGACAAT---TTCTTTCTCTCGATGC\n",
      "1 AAAGTTTGAATAGCTT------------\n",
      "1 AAGTTTGAATAGCTTC---------G-\n",
      "1 AAGTTTGAAT------TG-------AC\n",
      "1 AAGTTTGAA------------CGCAGC\n",
      "1 AATTTTGAATAGCTT------------\n",
      "1 AAG-TTGAA----------------GC\n",
      "1 AAGTTCGCA-A----------------\n",
      "1 AAAATTGAA----------------GC\n",
      "1 AAGTTTGGAGA---------------A\n",
      "1 ATGGCTGAATAGCTT------------\n",
      "1 AAAGTTGAA------------------\n",
      "1 AAGTTTGGAATAGCTTC-------AAG-\n",
      "1 AAG---GAA-AG-TTGAAT---TGAGT\n",
      "1 ----------AG---------------\n",
      "1 AAGTTTGAAGATGAGCTT------------\n",
      "1 AAGTTTGAATAGCTTC-------AAG-\n",
      "1 ----TCG-ATGGC----G-----ATGG\n",
      "1 AAGATTTGA-------------------\n",
      "1 AAGTTTGAATAGCTT---------AG-\n",
      "1 AAGTTTCGAATAGCTT------------\n",
      "1 AAGTCTGAA------------------\n",
      "1 AAGTTTGAA-A-------T-CCTTT--\n",
      "1 AAGTTTGAATAGCT-------------\n",
      "1 --GTTTGAATAGCTT------------\n",
      "1 AAGTTTGAA-ACGGAAACGGAAAT--CGAGGG\n",
      "1 TCGGTTGAA------------------\n",
      "1 --GTTCG-AGAG-TTCT-------T--\n",
      "1 AAGTTTGAA-ACGGAAACGGAAATTGTAGGG\n",
      "1 AAGTTTGAATAG---------CTCT--\n",
      "1 AAGTTTG-------TCAGGCACT----\n",
      "1 AA-ATTG--TAGCCACAG-----CTG-\n",
      "1 AAGTTTGAATGGCTT------------\n",
      "1 GGGTTTGAAT----------------C\n",
      "1 -----------GCTT---------T--\n",
      "1 AAGTTTGAATAGCT--TG-------G-\n",
      "1 A-----GA------------------C\n",
      "1 AAGTTTAAAGAG---------------\n",
      "1 AAGTTTGAAATAGGTT------------\n",
      "1 AAGTTTGAAAAG-------------AA\n",
      "1 ------------CCCCCGTCTGGTCCCAGCCCTTTAGAGAAGGG\n",
      "1 GAG------------------------\n",
      "1 AAGTTTG---------------TC-AA\n",
      "1 AAGTTTGA---------G------TG-\n",
      "1 AAGTTTGAATAGCTT-TGT--------\n",
      "1 AAGTT------GC-------------A\n",
      "1 AAGTTTGAA--------GT--------\n",
      "1 AAGTTTGA---------GTC-------\n",
      "1 AAGTTTG--TAGCTT------------\n",
      "1 AAGTTTG----G-----GT--------\n",
      "1 AAGTTTGATAAG---CTG-------G-\n",
      "1 AAGTTTG--------------------\n",
      "1 AGGTT----------------CTC---\n",
      "1 AAGTT----TAG---------------\n",
      "1 AAGTTTGAA-ACGGGAACGGAAAACAGAGGG\n",
      "1 ----------A----CGGTCC------\n",
      "1 AAGTTTGAATA-----AGTCC----G-\n",
      "1 AAGTTTGAATA----------------\n",
      "1 AAGTTTGAATAAGCT----------TG-\n",
      "1 AAGTTTGAAGCAGCT--TG-------G-\n",
      "1 AAGTTTGAA----------------GC\n",
      "1 AAGTTTGAA-ACGGAAACGG---AAATG-\n",
      "1 AAGTT-----------------TCTAA\n",
      "1 AAGTTTG--TAG---------------\n",
      "1 AAGTTTG----G-----GGGCC---G-\n",
      "1 AAGTTTGAAGAG-----GT--------\n",
      "1 AAGT------------------TCT--\n",
      "1 AAGT------------TG-------GC\n",
      "1 -AG---GAA--------GT---TCT-A\n",
      "1 AAGTTTG------------------GA\n",
      "1 AAGTTTGAA-A----------------\n",
      "1 AAGTTT---------------------\n",
      "1 AAGTTTGAATAG----------TCTG-\n",
      "1 AAGTT---CTAG---------------\n",
      "1 AAGTT-------CTTC-----------\n",
      "1 AAGTTTGAATAACTT-----------C\n",
      "1 AAGTTT----------------T-TAA\n",
      "1 AAGTT---CTAG-----GT-----TGG\n",
      "1 AAGTTTGAA-ACGGAAACGGAAAACAGAGGG\n",
      "1 AAGT---------TTC-GTC-----G-\n",
      "1 AAGTTCTAATAG----------T-TG-\n",
      "1 AAGTTTGAACA----------------\n",
      "1 AAGTT------GC------------GC\n",
      "1 AA---------------------CT--\n",
      "1 AAGTTTGAAGAGCT-------------\n",
      "1 AAGTTTGA-----------------G-\n",
      "1 AAGTTTGA--AGCT------------C\n",
      "1 ------------------------TGC\n",
      "1 --GCTTTGAATAGCTT------------\n",
      "1 AAAGTTCGAA------------------\n",
      "1 AAGTTCTGAACAG---------------\n",
      "1 ACGGTTGAA------------------\n",
      "1 AAGTTCGAATAGCTT------------\n",
      "1 AAGTTTG--TA---------------A\n",
      "1 AGGTTT---------------------\n",
      "1 AAGTTTGAAAA------------CT--\n",
      "1 AAGTTTGAA-A------------CGGA\n",
      "1 -AGTCTTTGAA--------G-------G-\n",
      "1 AAGTTTG------------------GC\n",
      "1 AAG----------------------G-\n",
      "1 AAGTTTGAAGAG-------------G-\n",
      "1 AAGTT----TA--------------GA\n",
      "1 AAGTTTG----G-----G------TG-\n",
      "1 AAGTTTGAATAG-----G------TG-\n",
      "1 AAGTTTGAA-ACGGAAACGGAAAGCAGAGC\n",
      "1 AAGTTTGAATAG----------TC-G-\n",
      "1 AAGTTTGAACA-------------AG-\n",
      "1 AAAGTGGATGAGCTT------------\n",
      "1 CAAGTTTGAACAGCTTC-----------\n",
      "1 AAGTTTGA--------TG-------G-\n",
      "1 GAG-TTGAA-----------------C\n",
      "1 AAGTTTG---AGAT------------C\n",
      "1 CAGTAGTTTGAA------------------\n",
      "1 AAGTT----------------CTAAG-\n",
      "1 AAGTTTCGATAG-TTC-----------\n",
      "1 AAGTTTGAA-ACGGAAACGGAAGCAGAGGG\n",
      "1 AAGTTTGA---------GT--------\n",
      "1 AAGTTTCCGGAAGAGAATAGCTT------------\n",
      "1 AAGTTTGAAGA----CGGT-----T--\n",
      "1 CCAAGTTTGAA------------------\n",
      "#### Processing NPM1\n",
      "2082 AAGTCTCTTTAAGAA------------\n",
      "1827 AAGTCTCTT------------------\n",
      "91 AAGTCTCTTTAA------AG----AA-\n",
      "53 AAGTCTCTTT----------------C\n",
      "39 AAGTCTCTTCACAG----TCAGTT-----\n",
      "30 ---------------------------\n",
      "29 AAATCTCTTTAAGAA------------\n",
      "27 AAGATGTCTT------------------\n",
      "16 AAGTCTCTTTAAGAAG-----------\n",
      "15 AAGTCTCTTT-----------------\n",
      "15 AAGTCTCTT----------G-------\n",
      "12 AAGTCTCTTT-AGAA------------\n",
      "10 ATGTC----------------TT----\n",
      "10 AAGTGTCTT------------------\n",
      "10 AAGTCTCTT-----------------C\n",
      "9 AAGATGTCTTTAAGAA------------\n",
      "9 AAGTGTCTTTAAGAA------------\n",
      "9 AAGTCTCTTCAAGAA------------\n",
      "9 AA---TCTT-----------CTT----\n",
      "7 AA-TCTTCTTTAAGAA------------\n",
      "7 AAGTCTCTTT-AGAGAA----------\n",
      "6 -AGTCTCTT------------------\n",
      "6 -AGTCTCTTTAAGAA------------\n",
      "5 A--TGTCTTTAAGAA------------\n",
      "5 ATGT-TCTT------------------\n",
      "5 AAGTCTCTTT-AGA----AG-------\n",
      "5 AAGTCTCTTTTAAGAA------------\n",
      "4 AAG--TCTTTAAGAA------------\n",
      "4 AA-TGTCTTTAAGAA------------\n",
      "4 -AGTGTCTT------------------\n",
      "4 AA-TGTCTT------------------\n",
      "4 AAGTCTCT------------------C\n",
      "4 AAGT-TCTT------------------\n",
      "4 AAAGTCTCTT------------------\n",
      "3 AATTGTCTT------------------\n",
      "3 AAGATGTCT-TTTAAGAA------------\n",
      "3 AAGTCTC-TTAAGAA------------\n",
      "3 AAGTCTTCTTTAAGAA------------\n",
      "3 AAGTCTCTTTAAGAAA-----------\n",
      "3 AAGTCTCTT----------------A-\n",
      "3 ATGT-TCTTTAAGAA------------\n",
      "3 AAGTCTCTTTAAGA----AG-------\n",
      "3 AAGTCTCTTT-AG-----AG-------\n",
      "3 AAGT-TCTTTAAGAA------------\n",
      "3 AAGTCTCTTTAAGA----AGT------\n",
      "3 AAGTCTCTTT--G------GT-----C\n",
      "3 AAAGTCTCTTTAAGAA------------\n",
      "3 AAGTCTCTTTAAG-----AG-------\n",
      "2 AAGTGTCTTT-----------------\n",
      "2 AATTGTCTTTAAGAA------------\n",
      "2 AA-TCTCTT------------------\n",
      "2 AAGTCTC--TAAGAA------------\n",
      "2 AAGTCTCTT----ACAA----------\n",
      "2 AAGTCTCTTT-AGA-------------\n",
      "2 -A-TGTCTTTAAGAA------------\n",
      "2 AAGTCTCTTTAACAAA-----------\n",
      "2 AAGTCTATT------------------\n",
      "2 -AGTGTCTTTAAGAA------------\n",
      "2 AAGTCTCTTT-AGAGA-----------\n",
      "2 AAGTCTCT------------TTTAA--\n",
      "2 AAGTGTCTTTTAAGAA------------\n",
      "2 AAGTCTCTT--------CAG-------\n",
      "2 AAGTATTTCTTTAAGAA------------\n",
      "2 AA-TCTTCTTAAGAA------------\n",
      "2 AAGTCTCTTTAA---------------\n",
      "2 AAGTCTCTT--AGAGAA----------\n",
      "2 -AGT---GTT-AGAGAA----------\n",
      "2 AAGT-ACTTTAAGAA------------\n",
      "2 AAGTCTCTTCAA---------------\n",
      "2 ------------GAAAAT-GTGTT----\n",
      "2 AAGTCTC-------------TTCAAAC\n",
      "2 AAGTCTCTTTAAGCAA-----------\n",
      "2 AAGTCTCTTTAAG--AACAG-------\n",
      "2 AAGTCTCTT---G------G-------\n",
      "2 AAGTCTCTTTGAAGAA------------\n",
      "2 AAATCTCTTTAA------AG----AA-\n",
      "2 AAGTCTCTTT---------G-----AC\n",
      "2 AA-T-TCTTTAA------AG----AA-\n",
      "2 AAGTCTCTTTAA------AG-------\n",
      "2 AAGTCTCTTTAAG------------AC\n",
      "2 AAGTCTCTTT-AGAAA-----------\n",
      "2 AAGT-GCTTT-----------------\n",
      "2 -AGT-TCTT------------------\n",
      "2 AAGTCTCTTTAAG-----AGT------\n",
      "2 AAAGTCTCTTTAA------AG----AA-\n",
      "2 AAGTCTCTTTAGG-------------C\n",
      "2 AAGTC-CTT------------------\n",
      "2 -AGTC----------------TT----\n",
      "1 AAGTTTCTTTAAGAA------------\n",
      "1 -AGT---------------GTTT---T\n",
      "1 AA-----TGTA---------TTT----\n",
      "1 -ATTTTCTT----------GT------\n",
      "1 AAGTCTCTTTAAG--------------\n",
      "1 AAG------------AAT-GTATT----\n",
      "1 AAGTCTCTTCAGG---AGAG-------\n",
      "1 AATTGTCTT--------CAG-------\n",
      "1 ATGTCT-TTTAAGAA------------\n",
      "1 ATGTCT-----------TAG-------\n",
      "1 AAGTCTCT---------TAG-------\n",
      "1 ------------GAAAATATTCT----\n",
      "1 -A-TCTCTTTAAGAA------------\n",
      "1 AAG--------------TA-TTT---G\n",
      "1 AAGTCTCTGTAAG------G-------\n",
      "1 -A-TCTCTTCAAGAA------------\n",
      "1 AAGTGTCTT--------TAG-------\n",
      "1 AAGTCTCTTT---AAAATAGTTTAAAC\n",
      "1 AAG--TCTTT---------G------C\n",
      "1 AAGTCTCTTCACTGA----AGCT-----\n",
      "1 AAGATCTCTT------------------\n",
      "1 AAGTGTCTTTAA---------------\n",
      "1 AAGTCTGTTCAAGAA------------\n",
      "1 AAGTCTCTTT-AGAGAAGA--------\n",
      "1 AAGTCTCTT-------ATGGT------\n",
      "1 AAG-CTCTCTAAG-----AG-----AC\n",
      "1 AATATTCTTCAAGAA------------\n",
      "1 AAGTCTCTTTAAGAACA-AG------C\n",
      "1 AAGTGTCTGTTGAGGA------------\n",
      "1 AAGTCTCTTGGAG-CAGTA--------\n",
      "1 AAGT-----------------------\n",
      "1 AAGTCTC--------------------\n",
      "1 AAGTCTCTTTAAGAAAATAGTTTAAAC\n",
      "1 AA-TCTTCTTTAAG-----------AAC\n",
      "1 ATGTAT-TTTAAGAA------------\n",
      "1 AA---TCT-----------GTT-----\n",
      "1 AGAGTGTCTT------------------\n",
      "1 AAGATGCTTTTTAAGAA------------\n",
      "1 AGAG--TGTTTAAGAA------------\n",
      "1 AAGT-ACT---------TAG-------\n",
      "1 AACCTTCT-TTTAAGAA------------\n",
      "1 ACAGTCTCTTTAAGAA------------\n",
      "1 -AG--TCTTT----------CTT----\n",
      "1 AAG--------------TACTTT----\n",
      "1 AAGTCTCTTTAAG--AATGGT------\n",
      "1 AAGTCTCTTTCAAGAA------------\n",
      "1 AAGTCTCTTCAAG-----AG-------\n",
      "1 AAATATATTTAAGAA------------\n",
      "1 AAG--------AG----TGACTT-----\n",
      "1 AAGTCTCTTTAAGAACA-AG-----AG\n",
      "1 AAGCTTCTTT----------------C\n",
      "1 AAGTCTCTTTAA----------TGAA-\n",
      "1 --------------------TACTTGAGC\n",
      "1 -A-TCTTCTTTAAGAA------------\n",
      "1 AAGGACTGTTTTTTCACAGTCAGTT-----\n",
      "1 AGAGTGTTTTTAAGAA------------\n",
      "1 AACTTTC-TTAAGAA------------\n",
      "1 AAGT---ATTAATGA------TTTA---\n",
      "1 AAATCTATTTAAGAA------------\n",
      "1 -AG--------AGAACCTTCTT-----\n",
      "1 AAGTCTCTTTGAGAA------------\n",
      "1 -AGT---------------G-TTGAAG\n",
      "1 AA-TCTTCTTGAACAA------------\n",
      "1 -----TCTT--------TAG-------\n",
      "1 -TGTCT-TTTAAGAA------------\n",
      "1 GAG--------------TACTT-----\n",
      "1 AAGTAT-TTTAAAAAAGTG--------\n",
      "1 AAGTATTTACTTTAAGAA------------\n",
      "1 AAGATGTTCTTAAGAA------------\n",
      "1 AAGTCTCTTGGAAGACAA----------\n",
      "1 AAGTCTCTTAAAG------G-------\n",
      "1 ATGAC---TTAAGAA------------\n",
      "1 AAAATGTATTTAAGAA------------\n",
      "1 AAGTGTCT------------TTT----\n",
      "1 AA---TGTT-----------TTT----\n",
      "1 AA-TCTCTTTTAAGAA------------\n",
      "1 AAGTCTCTTTAAGAACCAT-GTT-----\n",
      "1 AAATCT----------AT--TTTAAGA\n",
      "1 --------------------TTGTGCAGG\n",
      "1 AAAGCTCTTT-----------------\n",
      "1 AAGTCTCTTTA----------CTAA--\n",
      "1 AA-TCTTCTTGAAGAA------------\n",
      "1 -AGT-GCTTCAAGAA------------\n",
      "1 -----------AGAATATTCTTTGAAC\n",
      "1 -----------AGAATGTACTT-----\n",
      "1 AAAATATCTTTAAGAA------------\n",
      "1 AAGTCTCTTTA-----------CAAA-\n",
      "1 AAGATGTCTTCAAGAA------------\n",
      "1 ----------------ATTGCTTT----\n",
      "1 AAGTCTCTTT--GA----AG---CAAC\n",
      "1 AATGTCT---------------------\n",
      "1 ATGT---ATTAAGAA------------\n",
      "1 -AGTTGCTT----------G-------\n",
      "1 AAGGTTCTT--------CAG-------\n",
      "1 AAGTCT----------AT-G-------\n",
      "1 ATATGTCTT----ACAA------AAAC\n",
      "1 AAGTGTCTTT--GCA---AG-------\n",
      "1 AAGTCTCTTCAAG--------------\n",
      "1 -ATTTTCTTTAAGAA------------\n",
      "1 AAG-ATGTTCAAGAA------------\n",
      "1 -AG--TGTTTAAGAA------------\n",
      "1 ---------------AATATTCTTCAA-\n",
      "1 AAGTGTCTTTTAGAGAA----------\n",
      "1 AAAGTGTCTT------------------\n",
      "1 AAGTCTTCTTCAAGAA------------\n",
      "1 AAGTCTATGCAAG--ACTGGT------\n",
      "1 A--TCTCTTTAAGAA------------\n",
      "1 AAGTCTCTTT--------------AAC\n",
      "1 AAGTCTCTTTA-----------TAAA-\n",
      "1 AAGTCTCTT--CGAAAA-----CCAA-\n",
      "1 AAGTCTCTTCGCG---GTACATT----\n",
      "1 AGAATCTATTCTAAGAA------------\n",
      "1 AAGATGTCTT--AGAGAA----------\n",
      "1 ----------AAGA------TCTAA--\n",
      "1 AGATCTTCTTTAAGAA------------\n",
      "1 AAGTCTCTTT-----AATGG-----A-\n",
      "1 AAGTCTCTTT--GAA------------\n",
      "1 AAGTTGCTT------------------\n",
      "1 -AGTTTCTTTAAGAA------------\n",
      "1 -AGTGTC-TTAAGAA------------\n",
      "1 ATGT-GC-TTAAGAA------------\n",
      "1 AAGTCTCTT---GAAA-----------\n",
      "1 ATGTATCTTTAAGAA------------\n",
      "1 AAGCTGTCTTTAAGAA------------\n",
      "1 AAGTCTCTTTATGGAAA----------\n",
      "1 ------------GAAGAT-GCTTAAGAA\n",
      "1 AA-TGTATTTAAGAA------------\n",
      "1 AAGAATGTCTTTAAGAA------------\n",
      "1 -----------AGAAATATA--TTAAGAA\n",
      "1 AAGTGTCTTTAGAGAACA-AGT------\n",
      "1 -------TGTAAG----CAG------C\n",
      "1 AA-TCTCTTTAAGAA------------\n",
      "1 AAGTGTCTTTAAG------G------C\n",
      "1 AAG--TCTTCAAGAA------------\n",
      "1 GAGAATCTT-----------CTT----\n",
      "1 AA--CTATTGCAG----TCAGTT-----\n",
      "1 AAGATCTCTTTAAGAA------------\n",
      "1 AAGT---------------GTTCAA--\n",
      "1 AAGTC-C---AAGA--------T----\n",
      "1 AAGTCTCTTTA----------CTGAA-\n",
      "1 AAGTCTCTTTATGAA------------\n",
      "1 AAGTCTCT---------TACT------\n",
      "1 AAGTCTCTTCGAG--------------\n",
      "1 AAGTCTCTTTAAGAACGGAG----AA-\n",
      "1 AAAATGTCTTT-----------------\n",
      "1 AAGTCTCTTTACGAA------------\n",
      "1 --------------------GTTCCCAT\n",
      "1 AAGTCTCTTTCAGA-------------\n",
      "1 AAGTCTC------------G----AAC\n",
      "1 AAGTCTCTTTAAGAACAGAG-------\n",
      "1 -AGTGTCCTTTAAGAA------------\n",
      "1 -AG--------AGA------CTT----\n",
      "1 AAGTAT-TTT-AGAGAA----------\n",
      "1 AA-----TGTA----------TT---C\n",
      "1 A----TCTTCACAG----TGAGTT-----\n",
      "1 TAG-----TT-------TAG--TAAA-\n",
      "1 -AGTCT-TTT---------G-------\n",
      "1 AAATCTCTT------------------\n",
      "1 AAAGTC---------------TTT----\n",
      "1 AAGTCTCTTT-AGAGGAT---------\n",
      "1 AAGTCTCTTT--G------G---CAGC\n",
      "1 AAGTCTCTTT-AG-AAAT-GT------\n",
      "1 AAGTCT-TTTCTG--------------\n",
      "1 AAGTC------AG-----AGT------\n",
      "1 AA----------------AGTT-----\n",
      "1 AAAGTCCCTT---GAA------------\n",
      "1 AAGTCTCTTT--GA----AG-------\n",
      "1 -AGT-TCTTT---------G-------\n",
      "1 AAGTCTCTTT----AAAT-GCT-----\n",
      "1 AAGTCTCTTT---------G-------\n",
      "1 AAGTCTCTTTAAGAACA-AG-------\n",
      "1 AAGTCTCT-------------------\n",
      "1 AA-T-TTTTTAA------AG----AA-\n",
      "1 GAGTCTCTT------------------\n",
      "1 -AGT-TCTTTAAGAA------------\n",
      "1 AAGTCTCTT--------TAGT---AA-\n",
      "1 -AGTC---------------TTT----\n",
      "1 A------------AAAATCCTTT---G\n",
      "1 AAGTCTCTTTAAG--GCTAG-------\n",
      "1 AAGTCTCTTTAAGAA---CG-------\n",
      "1 -AG----TTT----------TTT----\n",
      "1 AAGTCTCTTTAAG-----------AAC\n",
      "1 CAAGTCTCTTTAAGAA------------\n",
      "1 AAGTCTCTTTAA-AGAA-AGTT-----\n",
      "1 GAGT-TCTTTAAGAA------------\n",
      "1 AAGT-TCT-----------GTT-----\n",
      "1 AAGATCTTTTT-------TCAGTT-----\n",
      "1 -AGTCTCTTTAAG----TAG---GCAC\n",
      "1 AAGTC-CTT-----------------C\n",
      "1 -AGT-TCTTT-----------------\n",
      "1 AAG----TTT-AGAGA-----------\n",
      "1 AA----CTT------------------\n",
      "1 AAGTCTCTT--------TAGTT---AC\n",
      "1 AAGTCTCTTT---------G------C\n",
      "1 ATAGTCT-----------TA---T----\n",
      "1 AAGTCTCTTTAAG-AAAT-GCT--AAC\n",
      "1 AAAAGTCCCTTTAAGAA------------\n",
      "1 AAGTC----------------------\n",
      "1 AAGTCTCTTT--GA----AG----AAT\n",
      "1 AAAT-TCTT------------------\n",
      "1 -AGTCTTCTT-----------------C\n",
      "1 AA--CTC---AAGAA------------\n",
      "1 AAG---------GAAACT--CTTCAA-\n",
      "1 -----------AGATAATTCTTTGAA-\n",
      "1 AAGTCTCTTTAAGA-------------\n",
      "1 AAGTCATCTTTAAGAA------------\n",
      "1 AAGCTTCTTTAAGAA------------\n",
      "1 -AGTCTCTTTAAGGAA-----------\n",
      "1 AAG--TCTTTCAG--------CT--AC\n",
      "1 ------CGT------AATCCTT-----\n",
      "1 AAGTC-------GAAGGCAG-------\n",
      "1 A-----CTT------------------\n",
      "1 AAGTCTCTTTTAAAGAA------------\n",
      "1 AAGTCTTCTT------------------\n",
      "1 AAGTCTCTTTGAAG--AACAGCT---AC\n",
      "1 AAGTCTCTT--CCACAGTCAGTT----G\n",
      "1 AGGTAATTCTTTAAGAA------------\n",
      "1 AAGTCTCTTTAGGA-------------\n",
      "1 ATGT-T-GTGAAGAA------------\n",
      "1 AAGTCTCTTTGAGA-------------\n",
      "1 A--TCTTTTTAAGAA------------\n",
      "1 AAGTCTCTTTAGAGAAAGGACCT-----\n",
      "1 AAGTCTCTTTAAG------G-------\n",
      "1 CAGT---TTTAAGAA------------\n",
      "1 GAGTC---TTCAG--------------\n",
      "1 AAGTCTCTTTGAAGAAA-----------\n",
      "1 AAGTCTCTTCACAG----TCAGTTCAAGG\n",
      "1 AAGTTCTTTTTAA------AG----AA-\n",
      "1 TAAT-TCTTTAAGAA------------\n",
      "1 -AGT-TCTTT-TG----TGGTT----G\n",
      "1 AAGTCTCTTTAA----------TAA--\n",
      "1 AAGTC-------G------G-------\n",
      "1 AAGCTTCTTTAA------AG----AA-\n",
      "1 AA-ACTCTTT-----------------\n",
      "1 AAGTCTCTTT-----AATAG-------\n",
      "1 AAGTCTCTTT-CGAA---GGGTTA---\n",
      "1 -AGTCTTCTTT-AG----TAG-------\n",
      "1 AAGTCTCTTTATG----CAG-------\n",
      "1 AAGTTCCTCTT------------------\n",
      "1 A-----CT------------TTT----\n",
      "1 A---------AAG---ATCCTT-----\n",
      "1 -AGT-T-TTGAAGAA------------\n",
      "1 AAGTCTCTTTAAG-AAGTAG-------\n",
      "1 AAG----TTGAAGAA------------\n",
      "1 AAGTCTCTTTAAG--AAT-GT-----C\n",
      "1 AAATCTCTTT-AGA----AG-------\n",
      "1 AAAT-TCTTTAAGAA------------\n",
      "1 AAGTC-CTTTAAGAA------------\n",
      "1 AAGTCTCTTT-----AAT-GCTT----\n",
      "1 AAGTCTCTTTAAGAACGGTGATTT----\n",
      "1 AAGTCTCTTT---AAAAT---------\n",
      "1 AAGTCTCTTCAAG------GGTT----\n",
      "1 AAGT-TCGGT--GA----AGGGT----\n",
      "1 AAGTCTCTTT-AGAA---AGTT-----\n",
      "1 AAGTCTCTTCAAG-AACTAG-------\n",
      "1 AAGTCTCTTT----AAA-AGT------\n",
      "1 AAG------------------TT----\n",
      "1 TAAGTCTCTTTAAGAA------------\n",
      "1 AAGTCTCTTT--GGAAGTAATGT----\n",
      "1 AAGTC-----AAGAA------------\n",
      "1 AAGTCTCTTTAGGAAGAAAGT---GAC\n",
      "1 AAGTC-C--GAAGAA------------\n",
      "1 GCAAGT-TCTTT-----------------\n",
      "1 AGATTC-GTTCAAGAA------------\n",
      "1 AAGTCTCTTT--GCAA------T----\n",
      "1 AAGTCTCTTTAAGAAA------T----\n",
      "1 AA---TCTT-----------TTT----\n",
      "1 AAGTCTCTTT-----AATGG-------\n",
      "1 -AGATTCTTTAAGAA------------\n",
      "1 AAGATATTCTT---GAGAACAG-------\n",
      "1 AAGATGTCTTT-AGAGAA----------\n",
      "1 AAGTATCTTTA------------CAAG\n",
      "1 AAGTACCTT------------------\n",
      "1 -AGTGTCTTCAAGGA------------\n",
      "1 AAGTCTGTTTAAGAA------------\n",
      "1 AAGTCTCTTCACAG----TGAGTT-----\n",
      "1 -----------AGAACCTTCTT-----\n",
      "1 ---TCTTTCTTTAAGAA------------\n",
      "1 -ATTTTCTTTTAGAA------------\n",
      "1 AAGTTCCTTTAGG---AT-GT------\n",
      "1 -AGT-T-TTTAAG---------TA---\n",
      "1 GAGTCTCTTCAAGAA------------\n",
      "1 AAGTCTCT---CGACAGTCAGTT-----\n",
      "1 AAGTCTCTTTAAGAA-----TTT----\n",
      "1 AAGTCGATCTT-----------------C\n",
      "1 AAG--TCTTTAAGA----AG-------\n",
      "1 AAGTTCTCTTGAAGA----AG---AAGG\n",
      "1 -AGTGTCT---------TAG-------\n",
      "1 AAGTCTCTTTAAGAACAAAT-GT------\n",
      "1 AAGTC-CTTGAAG------G-----AC\n",
      "1 AAACTCTCTT----CAAAT---------\n",
      "1 AAGTCTCTTTAA------AG----AAC\n",
      "1 AAGTCT------GAA------------\n",
      "1 AAGTCTCTTT---------GT-----C\n",
      "1 AAGTCTCTTTAAG-AACTAGT------\n",
      "1 AAGTCT--------------TTTA---\n",
      "1 AAGTCTCTTGAAGAA------------\n",
      "1 -----TGTTTAAGAA------------\n",
      "1 TAGTTTCTT------------------\n",
      "1 AAGTCTCTTT----------CTTAAGAA\n",
      "1 AAGTCTCTTT-AG--AACAGTT-----\n",
      "1 AAGTTCTTTTTAAGAA------------\n",
      "1 -AGTGTCTTT------ATA----AGAA\n",
      "1 AAGT-TCTTTA----------------\n",
      "1 AAGTCTCTT--AGA--------T----\n",
      "1 AAGTC-CTTGAAGAA------------\n",
      "1 AAGTCTC-TTAAGGAA-----------\n",
      "1 ----------ATGAAAAAGGTCTT----\n",
      "1 AAGTCTCTTGGAG----TGGCA--------\n",
      "1 -AGTCTCTTTAA------AG----AA-\n",
      "1 AA-TCTTCTTCAAGAA------------\n",
      "1 ----------------ATA--TTGAA-\n",
      "1 AAGT---ATTAAGAA------------\n",
      "1 AAGTGCATCT-----------G-------\n",
      "1 AAGTCTCTTCAAG------G-----AC\n",
      "1 AAGTCTCTTT--G------GCTCCAA-\n",
      "1 AA--C-C----------TA-TTT----\n",
      "1 AA-TGTCTTTAAGAAA-----------\n",
      "1 AAGTCTCT------------TTCAAAG\n",
      "1 AAATCTCTTT-AGAAA-----------\n",
      "1 AA---TCTTTAAGAA------------\n",
      "1 -AGTCTTCTT------------------\n",
      "1 -AGAATCTT-----------CTT----\n",
      "1 ------------GAAAAT-GTCTT----\n"
     ]
    }
   ],
   "source": [
    "#### TABULATE ENDOGENOUS LIGATION: sfGFP #####\n",
    "'''\n",
    "Tabulate ligation products for endogenous spligation of C-term sfGFP\n",
    "'''\n",
    "\n",
    "final_count_dict = {}\n",
    "full_span_dict = {}\n",
    "for alignment_id, bam_file in bams_1ref_splice.items():\n",
    "    ref_id = alignment_id.split(\"__\")[0]\n",
    "    ref = alignment_key.loc[\n",
    "        alignment_key.Sample_Description == ref_id\n",
    "    ].Reference_for_Alignment.values[0]\n",
    "\n",
    "    print(f\"#### Processing {ref_id}\")\n",
    "    bam = pysam.AlignmentFile(bam_file, \"rb\")\n",
    "\n",
    "    start = ref_roi_coords[\n",
    "        ref_roi_coords.Construct.str.startswith(ref)\n",
    "    ].Target1_0Based_Start.values[0]\n",
    "    end = start + 27\n",
    "\n",
    "    counts = Counter()\n",
    "\n",
    "    for read in bam.fetch(ref, start, end):\n",
    "        if read.is_unmapped:\n",
    "            continue\n",
    "\n",
    "        # FULL-SPAN REQUIREMENT\n",
    "        if read.reference_start > start or read.reference_end < end:\n",
    "            continue\n",
    "\n",
    "        aligned = []\n",
    "        ref_pos = read.reference_start\n",
    "        query_pos = 0\n",
    "\n",
    "        for op, length in read.cigartuples:\n",
    "\n",
    "            # M, =, X: matches\n",
    "            if op in (0, 7, 8):\n",
    "                for _ in range(length):\n",
    "                    if start <= ref_pos < end:\n",
    "                        aligned.append(read.query_sequence[query_pos])\n",
    "                    ref_pos += 1\n",
    "                    query_pos += 1\n",
    "\n",
    "            # I: insertions\n",
    "            elif op == 1:\n",
    "                if start <= ref_pos < end:\n",
    "                    aligned.append(read.query_sequence[query_pos:query_pos + length])\n",
    "                query_pos += length\n",
    "\n",
    "            # D: deletions\n",
    "            elif op in (2,3):\n",
    "                for _ in range(length):\n",
    "                    if start <= ref_pos < end:\n",
    "                        aligned.append(\"-\")\n",
    "                    ref_pos += 1\n",
    "\n",
    "            # Soft clip\n",
    "            elif op == 4:\n",
    "                query_pos += length\n",
    "\n",
    "            # Hard clip\n",
    "            elif op == 5:\n",
    "                pass\n",
    "\n",
    "        seq = \"\".join(aligned)\n",
    "\n",
    "        # Increment string count\n",
    "        if len(seq) > 0:\n",
    "            counts[seq] += 1\n",
    "\n",
    "    for seq, count in counts.most_common():\n",
    "        print(count, seq)\n",
    "\n",
    "    final_count_dict[ref_id] = counts\n",
    "    \n",
    "    # COUNT TOTAL FULL-SPAN READS\n",
    "    full_span_count = 0\n",
    "\n",
    "    for read in bam.fetch(ref, start, end):\n",
    "        if read.is_unmapped:\n",
    "            continue\n",
    "        if read.reference_start <= start and read.reference_end >= end:\n",
    "            full_span_count += 1\n",
    "            \n",
    "    full_span_dict[alignment_id.split(\"_\")[0]] = full_span_count\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 844,
   "id": "9bd30719",
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################### TABULATE PRIOR LIGATIONS ######################################\n",
    "sup_table_df = pd.read_excel(\"Spligation_Key_12162025.xlsx\", sheet_name=\"Sup_Table_ROIs\")\n",
    "final_count_dict_prior = {}\n",
    "full_span_dict_prior = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 845,
   "id": "de60c854",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#### Processing XIST_crNA_1\n",
      "\tGCCACTTGAACACTGCGACAGAACTGGATCCG\n",
      "#### Processing XIST_crNA_2\n",
      "\tTTGGACAACCTAACAAAGCACAGCCCGCCATG\n",
      "#### Processing Cis_1Cut\n",
      "\tGCGGATCAAGTTCTGTAGCAGTGTTCAAGTGG\n",
      "#### Processing Trans_Basic\n",
      "\tGCGGATCAAGTTCTGTCGCAGTGTTCAGATCAAGTTCTGTCGCAGTGTTCAAGTGG\n"
     ]
    }
   ],
   "source": [
    "#### TABULATE Prior: normal amplicon Ref ####\n",
    "'''\n",
    "Tabulate ligation products for prior XIST and reporter experiments\n",
    "'''\n",
    "hg38_processing = ['XIST_crNA_1', 'XIST_crNA_2']\n",
    "normal_processing = ['Cis_1Cut', 'Trans_Basic'] + hg38_processing\n",
    "\n",
    "normal_processing_df = sup_table_df.loc[sup_table_df.Analysis.isin(normal_processing)]\n",
    "\n",
    "final_count_dict_normal_prior = {}\n",
    "\n",
    "for row in normal_processing_df.itertuples():\n",
    "    alignment_id = row.Analysis\n",
    "    bam_file = row.BAM\n",
    "    chromo = row.chr\n",
    "    roi_str = row.ROI_Str\n",
    "\n",
    "    print(f\"#### Processing {alignment_id}\")\n",
    "    print(f\"\\t{roi_str}\")\n",
    "    bam = pysam.AlignmentFile(bam_file, \"rb\")\n",
    "\n",
    "    start = row.ROI_Start\n",
    "    end = start + len(roi_str)\n",
    "\n",
    "    counts = Counter()\n",
    "\n",
    "    for read in bam.fetch(chromo, start, end):\n",
    "        # SKIP UNMAPPED READS\n",
    "        if read.is_unmapped:\n",
    "            continue\n",
    "\n",
    "        # FULL-SPAN REQUIREMENT\n",
    "        if read.reference_start > start or read.reference_end < end:\n",
    "            continue\n",
    "\n",
    "        aligned = []\n",
    "        ref_pos = read.reference_start\n",
    "        query_pos = 0\n",
    "\n",
    "\n",
    "        for op, length in read.cigartuples:\n",
    "            # M, =, X: matches\n",
    "            if op in (0, 7, 8):\n",
    "                for i in range(length):\n",
    "                    if start <= ref_pos < end:\n",
    "                        aligned.append(read.query_sequence[query_pos])\n",
    "                    ref_pos += 1\n",
    "                    query_pos += 1\n",
    "\n",
    "            # I: insertions\n",
    "            elif op == 1:\n",
    "                if start <= ref_pos < end:\n",
    "                    aligned.append(read.query_sequence[query_pos:query_pos+length])\n",
    "                query_pos += length\n",
    "\n",
    "            # D: deletions\n",
    "            elif op in (2, 3):\n",
    "                for i in range(length):\n",
    "                    if start <= ref_pos < end:\n",
    "                        aligned.append(\"-\")\n",
    "                    ref_pos += 1\n",
    "\n",
    "            # Soft clip\n",
    "            elif op == 4:\n",
    "                query_pos += length\n",
    "\n",
    "            # Hard clip (ignore)\n",
    "            elif op == 5:\n",
    "                pass\n",
    "\n",
    "        seq = \"\".join(aligned)\n",
    "\n",
    "        # Increment count\n",
    "        if len(seq) > 0:\n",
    "            counts[seq] += 1\n",
    "\n",
    "    final_count_dict_prior[alignment_id] = counts\n",
    "    \n",
    "    \n",
    "    # COUNT TOTAL FULL-SPAN READS\n",
    "    full_span_count = 0\n",
    "\n",
    "    for read in bam.fetch(chromo, start, end):\n",
    "        if read.is_unmapped:\n",
    "            continue\n",
    "        if read.reference_start <= start and read.reference_end >= end:\n",
    "            full_span_count += 1\n",
    "            \n",
    "    full_span_dict_prior[alignment_id] = full_span_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 846,
   "id": "fdc109bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#### Processing Cis_2CutShort\n",
      "\tGCGGATCAAGTTCTGTCGCAGTGTTCAGATCAAGTTCTGTCGCAGTGTTCAAGTGG\n",
      "#### Processing Cis_2CutLong\n",
      "\tGCGGATCAAGTTCTGTCGCAGTGTTCAGATCAAGTTCTGTCGCAGTGTTCAAGTGG\n"
     ]
    }
   ],
   "source": [
    "#### PROCESS PRIOR: Ref with deletion ####\n",
    "'''\n",
    "Tabulate ligation products for 2-cut reporters with longer ROIs\n",
    "(as defined by \"end\" variable in line 22)\n",
    "'''\n",
    "del_processing = ['Cis_2CutShort', 'Cis_2CutLong']\n",
    "\n",
    "del_processing_df = sup_table_df.loc[sup_table_df.Analysis.isin(del_processing)]\n",
    "\n",
    "final_count_dict_del_prior = {}\n",
    "\n",
    "for row in del_processing_df.itertuples():\n",
    "    alignment_id = row.Analysis\n",
    "    bam_file = row.BAM\n",
    "    chromo = row.chr\n",
    "    roi_str = row.ROI_Str\n",
    "\n",
    "    print(f\"#### Processing {alignment_id}\")\n",
    "    print(f\"\\t{roi_str}\")\n",
    "    bam = pysam.AlignmentFile(bam_file, \"rb\")\n",
    "\n",
    "    start = row.ROI_Start\n",
    "    end = row.ROI_Stop+1\n",
    "\n",
    "    counts = Counter()\n",
    "\n",
    "    for read in bam.fetch(chromo, start, end):\n",
    "        # SKIP UNMAPPED READS\n",
    "        if read.is_unmapped:\n",
    "            continue\n",
    "\n",
    "        # FULL-SPAN REQUIREMENT\n",
    "        if read.reference_start > start or read.reference_end < end:\n",
    "            continue\n",
    "\n",
    "        aligned = []\n",
    "        ref_pos = read.reference_start\n",
    "        query_pos = 0\n",
    "\n",
    "\n",
    "        for op, length in read.cigartuples:\n",
    "            # M, =, X : Matches\n",
    "            if op in (0, 7, 8):\n",
    "                for i in range(length):\n",
    "                    if start <= ref_pos < end:\n",
    "                        aligned.append(read.query_sequence[query_pos])\n",
    "                    ref_pos += 1\n",
    "                    query_pos += 1\n",
    "\n",
    "            # I: insertions\n",
    "            elif op == 1:\n",
    "                if start <= ref_pos < end:\n",
    "                    aligned.append(read.query_sequence[query_pos:query_pos+length])\n",
    "                query_pos += length\n",
    "\n",
    "            # D: deletions\n",
    "            elif op in (2, 3):\n",
    "                for i in range(length):\n",
    "                    if start <= ref_pos < end:\n",
    "                        aligned.append(\"-\")\n",
    "                    ref_pos += 1\n",
    "\n",
    "            # Soft clip\n",
    "            elif op == 4:\n",
    "                query_pos += length\n",
    "\n",
    "            # Hard clip (ignore)\n",
    "            elif op == 5:\n",
    "                pass\n",
    "\n",
    "        seq = \"\".join(aligned)\n",
    "\n",
    "\n",
    "        if len(seq) > 0:\n",
    "            counts[seq] += 1\n",
    "\n",
    "    final_count_dict_prior[alignment_id] = counts\n",
    "    \n",
    "    \n",
    "    # COUNT TOTAL FULL-SPAN READS\n",
    "    full_span_count = 0\n",
    "\n",
    "    for read in bam.fetch(chromo, start, end):\n",
    "        if read.is_unmapped:\n",
    "            continue\n",
    "        if read.reference_start <= start and read.reference_end >= end:\n",
    "            full_span_count += 1\n",
    "            \n",
    "    full_span_dict_prior[alignment_id] = full_span_count\n",
    "\n",
    "final_count_dict.update(final_count_dict_prior)\n",
    "full_span_dict.update(full_span_dict_prior)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 847,
   "id": "e24a4be1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#### PROCESSING: LBR\n",
      "#### PROCESSING: NCL\n",
      "#### PROCESSING: NPM1\n",
      "#### PROCESSING: XIST_crNA_1\n",
      "#### PROCESSING: XIST_crNA_2\n",
      "#### PROCESSING: Cis_1Cut\n",
      "#### PROCESSING: Trans_Basic\n",
      "#### PROCESSING: Cis_2CutShort\n",
      "#### PROCESSING: Cis_2CutLong\n"
     ]
    }
   ],
   "source": [
    "#### SAVE ####\n",
    "'''\n",
    "Configure counts with display strings and save to Excel file\n",
    "'''\n",
    "\n",
    "# Adjust strings for minor alignment artifacts\n",
    "fix_strs = {'LBR':{'TACATCTACTAATGCTCTTC------T':'TACATCTACTAATGCTCTTCT------'},\n",
    "            'NCL':{},\n",
    "            'NPM1':{},\n",
    "            'Cis_1Cut':{'GCGGATCA------------------AAGTGG':'GCGGATCAA------------------AGTGG'},\n",
    "            'Cis_2CutShort':{}, \n",
    "            'Cis_2CutLong':{},\n",
    "            'Trans_Basic':{},\n",
    "            'XIST_crNA_1':{},\n",
    "            'XIST_crNA_2':{}\n",
    "            \n",
    "            \n",
    "}\n",
    "\n",
    "## COLLECT DATA\n",
    "dfs = {}\n",
    "for sample_id, data in final_count_dict.items():\n",
    "    \n",
    "    print(f\"#### PROCESSING: {sample_id}\")\n",
    "    \n",
    "    # FIX\n",
    "    fix_dict = fix_strs[sample_id]\n",
    "    for bad_str,new_str in fix_dict.items():\n",
    "        new_value = data[bad_str] + data[new_str]\n",
    "        data[new_str] = new_value\n",
    "    \n",
    "    # FILTER DICT FOR KNOWN LIGATION PRODUCTS\n",
    "    if 'XIST' in sample_id: # If Xist, convert to standardize target-site str format\n",
    "        temp_converter = xist_conversion_dict[sample_id]\n",
    "        data_clean = {temp_converter[k]:v for k,v in data.items() if k in temp_converter.keys()}\n",
    "        data_clean.update({k:0 for k in ligation_str_dict[sample_id] if k not in data_clean.keys()})\n",
    "    elif '2Cut' in sample_id:\n",
    "        data.update({k:0 for k in ligation_str_dict[sample_id] if k not in data.keys()})\n",
    "        temp_converter = large_del_conversion_dict[sample_id]\n",
    "        data_clean = {temp_converter[k]:v for k,v in data.items() if k in temp_converter.keys()}\n",
    "        data_clean.update({k:0 for k in ligation_str_dict[sample_id] if k not in data_clean.keys()})\n",
    "    elif \"Trans\" in sample_id:\n",
    "        data.update({k:0 for k in ligation_str_dict[sample_id] if k not in data.keys()})\n",
    "        temp_converter = trans_conversion_dict\n",
    "        data_clean = {temp_converter[k]:v for k,v in data.items() if k in temp_converter.keys()}\n",
    "    else:\n",
    "        data.update({k:0 for k in ligation_str_dict[sample_id] if k not in data.keys()})\n",
    "        data_clean = {k:v for k,v in data.items() if k in ligation_str_dict[sample_id]}\n",
    "    \n",
    "    \n",
    "    n_spanning_reads = full_span_dict[sample_id]\n",
    "\n",
    "    data_clean['Other_Reads'] = n_spanning_reads - sum(data_clean.values())\n",
    "    data_clean['Total_Reads'] = n_spanning_reads\n",
    "    \n",
    "    # Create DF\n",
    "    df = pd.DataFrame({'Ligation_Product':data_clean.keys(),\n",
    "                       'N_Reads':data_clean.values()\n",
    "                      })\n",
    "    \n",
    "    df['Percent'] = 100*df['N_Reads']/n_spanning_reads\n",
    "    dfs[sample_id] = df\n",
    "\n",
    "    \n",
    "## SAVE\n",
    "with pd.ExcelWriter(\"Junction_Counts_12222025.xlsx\", engine=\"openpyxl\") as writer:\n",
    "    for sample_id, df in dfs.items():\n",
    "        if df.empty:\n",
    "            continue\n",
    "\n",
    "        # Save to Excel sheet\n",
    "        sheet_name = sample_id[:31] # 31 chars max\n",
    "        df.to_excel(writer, sheet_name=sheet_name, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Test_Py_3_9",
   "language": "python",
   "name": "test_py_3_9"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
